src/TestSuite/TestCase.php

```diff
@@ -135,6 +161,29 @@ public function tearDown()
             Configure::write($this->_configure);
         }
         TableRegistry::clear();
+        self::_updateErrorLevel(true);
+    }
+
+    /**
+     * setUpBeforeClass
+     *
+     * @return void
+     */
+    public static function setUpBeforeClass()
+    {
+        parent::setUpBeforeClass();
+        self::_updateErrorLevel(false);
+    }
+
+    /**
+     * tearDownAfterClass
+     *
+     * @return void
+     */
+    public static function tearDownAfterClass()
+    {
+        parent::tearDownAfterClass();
+        self::_updateErrorLevel(true);
```

Comment:
Yes
For when we expect an exception (except deprecated) within  `deprecated()`

Reply:
Like `testAddInputTypeException` in above ,when you want to use `deprecated()`

Reply:
I thnk a bit
Probably we abe to remove this within `teardownAfterClass()` or `tearDown()` (just on of this) 
if you concern about performance
This is not related to using/not using `deprecated()`

Reply:
I think we should update the tests to use try/catch or have the deprecated helper catch exceptions, restore the error level and the rethrow the exception.

Reply:
We cant not try/catch for all tests at the moment 
And this way make disable all the error and is hard to update all tests
We can try/catch and rethrow within helper method (i tested this in local )
But this good for replaced tests not for current tests
So we must restore error level  on the `tearDown()` even by using good helper method for ensuring all tests(especially current tests ) no change the error level
What is problem by restore error level?
Performace?
If yes ,i can remove this method in `setUp()`
And if nedded in `tearDownAfterClass`

Reply:
If you want remove all `error_reporting()` from all tests
This is not possible IMO

Reply:
Do you think about removing all the calling  `error_reporting`

Reply:
I don't think we will be able to remove all calls to error_reporting() in the short term. I want to avoid adding more setup/teardown logic in the base class as it applies to every app/plugin as well. Instead we can apply setup/teardown to the tests that meed it now. Longer term we can fix those tests or start using deprecated().

Reply:
>I want to avoid adding more setup/teardown logic in the base class as it applies to every app/plugin as well.

```php
+        $needle = 'Cake\Test\TestCase';
+        $class = get_called_class();
+        if (substr($class, 0, 18) !== $needle) {
+            return;
+        }
```

Reply:
I'm not sure that is something we should be doing. I agree it will work, but I would rather take the time to fix the tests now.